### ISSUE TITLE
Difference between loss.backward() and model_engine.backward(loss) ?

### ISSUE THREAD

**Issue by rsn870** [microsoft/DeepSpeed#329]
Hi , 

I have tried out both loss.backward() and model_engine.backward(loss) for my code. There are several subtle differences that I have observed , for one retain_graph = True does not work for model_engine.backward(loss) . This is creating a problem since buffers are not being retained every time I run the code for some reason. 

Please look into this if you could.

**Comment by ShadenSmith** [comment]
> Hi , if one does loss.backward() instead of model_engine.backward(loss), so it will be wrong because gradients will not be averaged and therefore at every GPU, the weights will differ eventually leading us to train N models on N GPUs with batch_size/N samples per iteration.

Correct, you need `model_engine.backward()` to reduce gradients and perform other DeepSpeed bookkeeping.

> If the model_engine() is based on DDP() for example I think computed loss should be synchronized among all nodes ?

It is not based on DDP directly, so without `model_engine.backward()` you don't get synchronization.

> Also please tell me if there is any workaround for retain_graph=True for now?

Not currently, DeepSpeed needs to propagate additional arguments to our [backward interface](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/pt/deepspeed_light.py#L736). I self-assigned but will not be able to get to it until hopefully next week. We are very welcoming of PRs if you would like to contribute :-).  I would be happy to help guide through the process.

**Comment by kracwarlock** [comment]
@ShadenSmith it would be great if we can add `retain_graph` and other arguments.

**Comment by asit2898** [comment]
@ShadenSmith is this issue still open? If yes can I pick this up?

**Comment by ShadenSmith** [comment]
@asit2898 yes it is, apologies for us not getting to it quickly. PRs are very much welcome.

**Comment by asit2898** [comment]
Thanks a lot @ShadenSmith, I'll work on this issue. This would be my first commit, so any pointers would be truly helpful!

**Comment by LIZHICHAOUNICORN** [comment]
does any update?

**Comment by bhushan-borole** [comment]
Hey @ShadenSmith 
Is this issue open to work?
I believe @ncilfone has submitted a PR for this issue.

**Comment by buttercutter** [comment]
@ncilfone May I know why [retain_graph](https://github.com/microsoft/DeepSpeed/issues/2029#issuecomment-1183812888) still does not work for me ?
